plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs

```diff
@@ -326,14 +326,50 @@ sizeIsAcceptable = \case
   Delay _ t  -> sizeIsAcceptable t
 
 isPure :: Term name uni fun a -> Bool
-isPure = go
+isPure = go True
     where
-        go = \case
-            Var {}      -> True
+        -- See Note [delayAndVarIsPure]
+        go delayAndVarIsPure = \case
+            Var {}                        -> delayAndVarIsPure
             -- These are syntactically values that won't reduce further
-            LamAbs {}   -> True
-            Constant {} -> True
-            Delay {}    -> True
-
+            LamAbs {}                     -> True
+            Constant {}                   -> True
+            Delay _ body                  -> delayAndVarIsPure || go delayAndVarIsPure body
+            -- This case is not needed in PIR's `isPure`, because PIR's beta-reduction pass
+            -- turns terms like this into `Let` bindings.
+            Apply _ (LamAbs _ _ body) arg -> go delayAndVarIsPure arg && go delayAndVarIsPure body
+            Force _ body                  -> go False body
             -- See Note [Differences from PIR inliner] 5
-            _           -> False
+            Builtin{}                     -> True
```

Comment:
Hm, I thought in PIR we check if the specific builtin is underapplied, in which case it's pure if all of its arguments are pure and type/term argument interleaving is correct. Do I misremember or do we not want to handle this case here for some reason?

Reply:
`Note [Differences from PIR inliner]` explains why we don't do that. I actually tried it, and it didn't improve any of the existing test cases while adding a lot of lines of code.